plugins/query-resource-limit-group/src/main/java/org/opensearch/plugin/resource_limit_group/service/ResourceLimitGroupPersistenceService.java

```diff
@@ -0,0 +1,460 @@
+        double value = 0.0;
+        List<ResourceLimit> resourceLimits = resourceLimitGroup.getResourceLimits();
+        for (ResourceLimit resourceLimit : resourceLimits) {
+            if (resourceLimit.getResourceName().equals(resourceName)) {
+                value = resourceLimit.getValue();
+                break;
+            }
+        }
+        return value;
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param resourceLimitGroup - the resource limit group we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveResourceLimitGroupInClusterState(final ResourceLimitGroup resourceLimitGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = resourceLimitGroup.getName();
+        if (metadata.resourceLimitGroups().containsKey(groupName)) {
+            logger.warn("Resource Limit Group with name {} already exists. Not creating a new one.", groupName);
+            throw new RuntimeException("Resource Limit Group with name " + groupName + " already exists. Not creating a new one.");
+        }
+        final List<ResourceLimitGroup> previousGroups = new ArrayList<>(metadata.resourceLimitGroups().values());
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        String resourceNameWithThresholdExceeded = "";
+        for (ResourceLimit resourceLimit : resourceLimitGroup.getResourceLimits()) {
+            String resourceName = resourceLimit.getResourceName();
+            double existingUsage = calculateExistingUsage(resourceName, previousGroups, groupName);
+            double newGroupUsage = getResourceLimitValue(resourceName, resourceLimitGroup);
+            inflightResourceLimitValues.computeIfAbsent(resourceName, k -> new DoubleAdder()).add(newGroupUsage);
+            double totalUsage = existingUsage + inflightResourceLimitValues.get(resourceName).doubleValue();
+            if (totalUsage > 1) {
+                resourceNameWithThresholdExceeded = resourceName;
+            }
+        }
+        restoreInflightValues(resourceLimitGroup);
+        if (!resourceNameWithThresholdExceeded.isEmpty()) {
+            logger.error("Total resource allocation for {} will go above the max limit of 1.0", resourceNameWithThresholdExceeded);
+            throw new RuntimeException(
+                "Total resource allocation for " + resourceNameWithThresholdExceeded + " will go above the max limit of 1.0"
+            );
+        }
+
+        // check if group count exceed max
+        boolean groupCountExceeded = inflightCreateResourceLimitGroupRequestCount.incrementAndGet() + previousGroups
+            .size() > maxResourceLimitGroupCount;
+        inflightCreateResourceLimitGroupRequestCount.decrementAndGet();
```

Comment:
This should be in the terminal hooks of ClusterUpdateTasks. Here this will be logically invalid, since you should not decrement this until the task is failed or succeeded